### PR TITLE
feat: add upstream directive

### LINKED ISSUE #473: Move `@server.upstream` to `@upstream`
**Before**

```graphql
schema @server(port: 8080, upstream: {user_agent: "graphql/tailcall"}) {
  query: Query  
}
```

**After**

```graphql
schema @server(port: 8080) @upstream(user_agent: "graphql/tailcall") {
  query: Query  
}
```

This restructure makes the DSL look a lot cleaner and keeps the upstream configurations out of the general server configurations.

### PR DESCRIPTION
**Pull request by Yougigun**
**Summary:**  
_Briefly describe the changes made in this PR._
1. add upstream directive 
2. still keep the upstream as one field in Server type
**Issue Reference(s):**  
Fixes #473 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh` to address and fix linting issues.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the documentation (https://tailcall.run) accordingly. 
https://github.com/tailcallhq/tailcallhq.github.io/pull/21
- [x] I have performed a self-review of my own code.
- [x] Update `.tailcall.graphql`

### CODE REVIEW

File: src/config/from_document.rs [hunk 1]
```diff
@@ -329,3 +336,89 @@ impl TryFrom<ServiceDocument> for Config {
     from_document(value)
   }
 }
+
+#[cfg(test)]
+mod tests {
```
Comment (tusharmath) [inline src/config/from_document.rs:341]: You can instead write tests using graphql files. Checkout the integration `/tests` folder.

Comment (Yougigun) [reply src/config/from_document.rs:341]: Done. Realized that there are some new changes in integration tests. After fixing them, it is all fine and covers this feature. Help review, thanks!

File: src/config/from_document.rs [hunk 2]
```diff
@@ -329,3 +336,89 @@ impl TryFrom<ServiceDocument> for Config {
...
+  use async_graphql::Pos;
+  use async_graphql_value::ConstValue;
+  use indexmap::IndexMap;
+  use serde_json::Number;
+
+  use super::*;
+  use crate::config::Proxy;
+
+  #[test]
+  fn test_upstream_directive_in_server() {
```
Comment (tusharmath) [inline src/config/from_document.rs:352]: What are we testing here that's not captured in the other tests?

Comment (Yougigun) [reply src/config/from_document.rs:352]: I think the integration test does cover this part.  It is merely a unit test for a certain function.

Comment (tusharmath) [reply src/config/from_document.rs:352]: I think we can drop them and just keep the integration tests for this purpose.

Comment (Yougigun) [reply src/config/from_document.rs:352]: okay!

File: src/config/server.rs [hunk 3]
```diff
@@ -16,7 +16,7 @@ pub struct Server {
   pub enable_response_validation: Option<bool>,
   pub global_response_timeout: Option<i64>,
   pub port: Option<u16>,
-  #[serde(default, skip_serializing_if = "is_default")]
+  #[serde(skip)]
   pub upstream: Upstream,
```
Comment (tusharmath) [inline src/config/server.rs:20]: I think we can just delete this field.

File: src/config/into_document.rs [hunk 4]
```diff
@@ -12,7 +12,10 @@ fn config_document(config: &Config) -> ServiceDocument {
   let mut definitions = Vec::new();
   let schema_definition = SchemaDefinition {
     extend: false,
-    directives: vec![pos(config.server.to_directive("server".to_string()))],
+    directives: vec![
+      pos(config.server.to_directive("server".to_string())),
+      pos(config.server.upstream.to_directive("upstream".to_string())),
+    ],
```
Comment (tusharmath) [inline src/config/into_document.rs:18]: 👍

File: src/config/server.rs (hunk 3 above)
Comment (Yougigun) [reply src/config/server.rs:20]: I was thinking to not touch the logic of request handling upstrem in Server struct. So I only filled in the upstream field of Server from new upstream directive. If we are gonna delete the upstream filed then I will move to Config’filed thay is same level with Server field in config. then also refactor the upstream request to this new upstream filed. what do yoy think ?

Comment (tusharmath) [reply src/config/server.rs:20]: Yeah, I think that's the way we should go about this.

Comment (Yougigun) [reply src/config/server.rs]: @tusharmath 
Done! Help review, thanks.

File: src/blueprint/blueprint.rs [hunk 5]
```diff
@@ -19,6 +20,7 @@ pub struct Blueprint {
   pub definitions: Vec<Definition>,
   pub schema: SchemaDefinition,
   pub server: Server,
+  pub upstream: Upstream,
```
Comment (Yougigun) [inline src/blueprint/blueprint.rs:23]: @aaditmshah 
While I was adding a new upstream directive and refactoring the upstream request, I encountered a conflict. During the resolution, I saw your comment about moving the upstream from the server to the blueprint. I went ahead and moved it accordingly. Could you please check to ensure everything is in order?

File: tests/graphql/test-http-headers.graphql [hunk 6]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {baseURL: "http://localhost:4000"}) {
+schema @server(port: 4001) @upstream(baseURL: "http://localhost:4000") {
```
Comment (tusharmath) [inline tests/graphql/test-http-headers.graphql:2]:
```suggestion
schema @server @upstream(baseURL: "http://localhost:4000") {
```

do we need to set this port?

**Comment by Yougigun** [comment]
@tusharmath Is anything block this pr, otherwise could you help to merge before next conflict occurs! 😅

File: src/blueprint/server.rs [hunk 7]
```diff
@@ -19,8 +19,6 @@ pub struct Server {
   pub global_response_timeout: i64,
   pub port: u16,
   pub hostname: IpAddr,
-  // TODO: add blueprint::Server::Upstream
-  pub upstream: crate::config::Upstream,
```
Comment (tusharmath) [inline src/blueprint/server.rs:23]: 👍

File: src/blueprint/server.rs [hunk 8]
```diff
@@ -42,13 +40,12 @@ impl Server {
   }
 }
 
-impl TryFrom<crate::config::Server> for Server {
+impl TryFrom<crate::config::Config> for Server {
```
Comment (tusharmath) [inline src/blueprint/server.rs:43]: shouldn't Server be created from Server?

File: src/blueprint/server.rs [hunk 9]
```diff
@@ -108,29 +105,18 @@ fn handle_response_headers(resp_headers: BTreeMap<String, String>) -> Valid<Head
...
 
-fn handle_base_url(base_url: Option<String>) -> Valid<Option<String>, String> {
-  let base_url = if let Some(base_url) = base_url {
-    Valid::Ok(reqwest::Url::parse(base_url.as_str()).map_err(|e| ValidationError::new(e.to_string()))?)?;
-    Some(base_url)
-  } else {
-    None
-  };
-  Ok(base_url)
-}
```
Comment (tusharmath) [inline src/blueprint/server.rs:119]: We have deleted this validation I think. Can you add it back with a unit test?

File: src/http/client.rs [hunk 10]
```diff
@@ -28,16 +27,14 @@ pub struct DefaultHttpClient {
 
 impl Default for DefaultHttpClient {
   fn default() -> Self {
-    let server = config::Server::default();
+    let upstream = config::Upstream::default();
     //TODO: default is used only in tests. Drop default and move it to test.
-    DefaultHttpClient::new(Server::try_from(server).unwrap())
+    DefaultHttpClient::new(upstream)
```
Comment (tusharmath) [inline src/http/client.rs:32]: brilliant!

File: src/http/request_context.rs [hunk 11]
```diff
@@ -5,28 +5,29 @@ use hyper::HeaderMap;
 
 use super::{DefaultHttpClient, Response, ServerContext};
 use crate::blueprint::Server;
-use crate::config;
+use crate::config::{self, Upstream};
 
 #[derive(Setters)]
 pub struct RequestContext {
   pub http_client: DefaultHttpClient,
   pub server: Server,
```
Comment (tusharmath) [inline src/http/request_context.rs:13]: do we even need `server`? the idea of moving upstream out was to actually make the client decoupled with the server settings.

File: tests/graphql_spec.rs [hunk 12]
```diff
@@ -198,7 +198,7 @@ async fn test_execution() -> std::io::Result<()> {
   let specs = GraphQLSpec::cargo_read("tests/graphql/passed");
 
   for spec in specs? {
-    let mut config = Config::from_sdl(&spec.server_sdl[0]).unwrap();
+    let mut config = Config::from_sdl(dbg!(&spec.server_sdl[0])).unwrap();
```
Comment (tusharmath) [inline tests/graphql_spec.rs:201]:
```suggestion
    let mut config = Config::from_sdl(&spec.server_sdl[0]).unwrap();
```

File: tests/graphql/test-http-baseurl.graphql [hunk 13]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {baseURL: "http://abc.com"}) {
+schema @server @upstream(baseURL: "http://abc.com") {
```
Comment (tusharmath) [inline tests/graphql/test-http-baseurl.graphql:2]: Can we not live without `@server`, is it a required directive?

File: src/blueprint/server.rs (hunk 8 above)
Comment (Yougigun) [reply src/blueprint/server.rs]: oh! yes we can after move the upstream out of server. will do.

File: src/blueprint/server.rs (hunk 9 above)
Comment (Yougigun) [reply src/blueprint/server.rs:119]: oops. okay, i will add it back and invoke it in the config_blueprint  process. according to unit test, do you mean for handle_base_url function or handle_base_url(the function call handle_base_url)

File: tests/graphql/test-http-headers.graphql (hunk 6 above)
Comment (Yougigun) [reply tests/graphql/test-http-headers.graphql:2]: removed

File: tests/graphql/test-http-baseurl.graphql (hunk 13 above)
Comment (Yougigun) [reply tests/graphql/test-http-baseurl.graphql:2]: I am not quite sure about it. In the codebase and integration test, it seems required so far.

File: src/http/request_context.rs (hunk 11 above)
Comment (Yougigun) [reply src/http/request_context.rs:13]: I tried to remove the server field from RequestContext. But there are some values needed in the Server.
e.g. 
https://github.com/tailcallhq/tailcall/blob/main/src/lambda/expression.rs#L124
https://github.com/tailcallhq/tailcall/blob/main/src/lambda/evaluation_context.rs#L61

What do you think? Move these fields upstream. I'm just afraid to incur extra conflict.

File: src/blueprint/from_config.rs [hunk 14]
```diff
@@ -32,17 +32,29 @@ pub fn config_blueprint(config: &Config) -> Valid<Blueprint> {
...
+fn handle_base_url(base_url: Option<String>) -> Valid<Option<String>> {
+  let base_url = if let Some(base_url) = base_url {
+    Valid::Ok(reqwest::Url::parse(base_url.as_str()).map_err(|e| ValidationError::new(e.to_string()))?)?;
+    Some(base_url)
+  } else {
+    None
+  };
+  Ok(base_url)
+}
+
```
Comment (Yougigun) [inline src/blueprint/from_config.rs:51]: I added it back here.

File: src/blueprint/server.rs (hunk 8 above)
Comment (Yougigun) [reply src/blueprint/server.rs]: done

File: src/blueprint/server.rs (hunk 9 above)
Comment (Yougigun) [reply src/blueprint/server.rs:119]: here 
https://github.com/tailcallhq/tailcall/pull/479/files/7869161c5912cf7b0d6fe32f77841e58922084b5#diff-0934da3dca97565bf71c4fd107919b495a6a971c292009709dfc4756760ac27eR35-R51

File: src/blueprint/from_config.rs [hunk 15]
```diff
@@ -32,17 +32,29 @@ pub fn config_blueprint(config: &Config) -> Valid<Blueprint> {
...
   let server: Server = Server::try_from(config.server.clone())?;
-  let blueprint = Blueprint { schema, definitions, server };
+  let mut upstream = config.upstream.clone();
+  upstream.base_url = handle_base_url(upstream.base_url)?;
+  let blueprint = Blueprint { schema, definitions, server, upstream };
   let blueprint = apply_batching(blueprint);
   Ok(super::compress::compress(blueprint))
 }
 
+fn handle_base_url(base_url: Option<String>) -> Valid<Option<String>> {
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:42]:
```suggestion
fn handle_base_url(base_url: Option<String>) -> Valid<()> {
```
This function doesn't need to return anything, just validate.

File: src/http/request_context.rs (hunk 11 above)
Comment (tusharmath) [reply src/http/request_context.rs:13]: I think we can do this in a separate issue. Would you mind creating one?

File: tests/graphql/test-inline.graphql [hunk 16]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {baseURL: "http://jsonplacheholder.typicode.com"}) {
+schema @server(port: 4000) @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```
Comment (tusharmath) [inline tests/graphql/test-inline.graphql:2]:
```suggestion
schema @server @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```

File: tests/graphql/test-interface.graphql [hunk 17]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {baseURL: "http://jsonplacheholder.typicode.com"}) {
+schema @server(port: 4000) @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```
Comment (tusharmath) [inline tests/graphql/test-interface.graphql:2]:
```suggestion
schema @server @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```

File: tests/graphql/test-unsafe.graphql [hunk 18]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {baseURL: "http://jsonplacheholder.typicode.com"}) {
+schema @server(port: 4000) @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```
Comment (tusharmath) [inline tests/graphql/test-unsafe.graphql:2]:
```suggestion
schema @server @upstream(baseURL: "http://jsonplacheholder.typicode.com") {
```

File: tests/graphql/test-upstream.graphql [hunk 19]
```diff
@@ -1,5 +1,5 @@
 #> server-sdl
-schema @server(upstream: {proxy: {url: "http://localhost:8080"}}) {
+schema @server(port: 8001) @upstream(proxy: {url: "http://localhost:8085"}) {
```
Comment (tusharmath) [inline tests/graphql/test-upstream.graphql:2]:
```suggestion
schema @server @upstream(proxy: {url: "http://localhost:8085"}) {
```

File: src/blueprint/from_config.rs [hunk 20]
```diff
@@ -32,17 +32,29 @@ pub fn config_blueprint(config: &Config) -> Valid<Blueprint> {
...
+  let base_url = if let Some(base_url) = base_url {
+    Valid::Ok(reqwest::Url::parse(base_url.as_str()).map_err(|e| ValidationError::new(e.to_string()))?)?;
+    Some(base_url)
+  } else {
+    None
+  };
+  Ok(base_url)
+}
```
Comment (tusharmath) [inline src/blueprint/from_config.rs:50]:
```suggestion
fn handle_base_url(base_url: Option<String>) -> Valid<Option<String>> {
  if let Some(base_url) = base_url {
    reqwest::Url::parse(base_url.as_str())
      .map_err(|e| ValidationError::new(e.to_string()));
  }
  Ok(())
}
```

**Comment by tusharmath** [comment]
@Yougigun I think this is close to completion, just a few comments for which I have already added suggestions. just need to commit and run lint and we are good to go!

**Comment by tusharmath** [comment]
There are a few quick `todos` in the description, please have a look at them also 🙏

File: src/blueprint/from_config.rs (hunk 20 above)
Comment (Yougigun) [reply src/blueprint/from_config.rs:50]: seems Ok(()) not match the return type

Comment (Yougigun) [reply src/blueprint/from_config.rs]: done. please check

File: src/blueprint/from_config.rs [hunk 21]
```diff
@@ -32,17 +32,29 @@ pub fn config_blueprint(config: &Config) -> Valid<Blueprint> {
   let schema = to_schema(config)?;
   let definitions = to_definitions(config, output_types, input_types)?;
   let server: Server = Server::try_from(config.server.clone())?;
-  let blueprint = Blueprint { schema, definitions, server };
+  let mut upstream = config.upstream.clone();
+  upstream.base_url = handle_base_url(upstream.base_url)?;
+  let blueprint = Blueprint { schema, definitions, server, upstream };
```
[the remark this answers is not in the document]
Comment (Yougigun) [reply src/blueprint/from_config.rs]: done

File: src/http/request_context.rs (hunk 11 above)
Comment (Yougigun) [reply src/http/request_context.rs:13]: sure! #504

File: examples/.tailcallrc.graphql [hunk 22]
```diff
@@ -8,10 +8,26 @@ directive @server(
   globalResponseTimeout: Int
   port: Int
   vars: [KeyValue]
-  upstream: Upstream
   responseHeaders: [KeyValue]
   hostname: String
 ) on SCHEMA
+directive @Upstream(
```
Comment (tusharmath) [inline examples/.tailcallrc.graphql:14]:
```suggestion
directive @upstream(
```

**Comment by tusharmath** [comment]
@Yougigun Thank you 🙏 This was an unexpectedly large change.

**Pull request merged**